### PR TITLE
[2.10] Properly format windows env-vars, redeploy windows plans if they fail

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
## Problem

Windows provisioning currently suffers from instability to due to multiple issues. For instance, environment variables utilized by the `system-agent-installer-rke2` image are not properly formatted by Rancher and as such cannot be used. This results in `rke2` being stopped and reinstalled unnecessarily, which has the potential to result in a plan failure. Additionally, there are cases where non-Rancher managed Windows services (as as vSphere `vmtools`) may temporarily prevent rancher-wins from properly applying plans. In these cases, Rancher should automatically reattempt the plan to ensure that it is successfully applied. 

## Solution

Implement a new helper function for properly formatting windows environment variables, `FormatWindowsEnvVar`.  `FormatWindowsEnvVar` will convert a `corev1.EnvVar` into the proper string representation for use in either a PowerShell script (like the rancher-wins install script) or a planner variable (which is later passed to an `exec.Command`). 

Deliver two new environment variables when creating the `system-agent-installer-rke2` install plan, `WINS_RESTART_STAMP` and `INSTALL_RKE2_VERSION`. These environment variables will be utilized by the `system-agent-installer-rke2` to ensure that the `rke2` service is only reinstalled or restarted when needed. Other preexisting environment variables have also been updated to use `FormatWindowsEnvVar`. 

Update the Rancher planner to respect the `max-failures` and `failure-threshold` plan fields. Previously, `-1` and `1` were always passed, which prevents plans from being reattempted after a failure. Now, Rancher will attempt to reapply the windows `system-agent-installer-rke2` plan up to 5 times before declaring a plan failure. All Linux plans continue to use `-1` and `1`. Additionally, a new field has been added to the `OneTimeInstruction` struct, `ResetFailureCountOnServiceRestart`. When this field is set to true, the system-agent will clear the `failure-count` and `failed-checksum` values on the plan when force applying the instruction after a restart. This ensures that the plans `max-failure` value will be respected if the node or system-agent is restarted. 

**Note that an associated `system-agent` change and `rancher-wins` bump is needed to fully address this issue, so this PR shouldn't be merged until that occurs**

 
### Manual Testing

I have built a custom version of Rancher with these changes and tested them against modified versions of `rancher-wins` and the `system-agent-installer-rke2` image. I was able to confirm that the required environment variables are now properly formatted and that the plan will be reattempted up to five times. I also tested these changes against vSphere and saw improved stability post vm reboot

### Automated Testing

* Test types added/modified:
    * Unit
Summary:

I've added unit testing for the new `FormatWindowsEnvVar` function to ensure that it works as expected. Existing unit tests covering the relevant environment variables have also been updated to cover these changes.

## QA Testing Considerations

While some aspects of this PR will be testable, such as reattempting plans multiple times, other aspects relating to the new environment variables will only be testable using recent versions of the `system-agent-installer-rke2` image (i.e. recent versions of rke2). 
 
### Regressions Considerations

Existing / newly added automated tests that provide evidence there are no regressions:

### CODE REVIEW

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/capr/planner/instructions.go
Comment (HarrisonWAffel) [reply pkg/capr/planner/instructions.go:45]: Are you referring to line 41 compared to line 42? If so, thats intentional. I could comment that line out if we want, it is related to the rke2 data dir feature that went out a few releases ago. 

Unfortunately that implementation hit the same formatting issues that all other Windows env vars hit, so it has never worked. The RFC for that data-dir feature did not consider Windows, and there are several other places that need to be updated to have everything work as expected. I'd rather not touch anything related to the feature until we can create a follow-up RFC for windows support, so I did not touch that env var.

Comment (jiaqiluo) [reply pkg/capr/planner/instructions.go:45]: Yes,  that is what I am referring to. Sorry that I did not make it clear. 

Thanks for the explanation. Just to confirm my understanding: the `env` variable in this function is used in Rancher's Plan (not in the PowerShell script), and each element should follow the format `Name=Var`. However, the `date-dir` environment variable generated on line 41 is incorrectly formatted, but we leave it as is. 

Does it sound right to you?

Comment (HarrisonWAffel) [reply pkg/capr/planner/instructions.go:45]: Correct. Since the system-agents applyinator will [execute plans using `cmd.Exec`](https://github.com/rancher/system-agent/blob/7ad21ffbe9273d33d28fbf17249e949fdc6d8610/pkg/applyinator/applyinator.go#L504-L533), which automatically prepends `$env:` to variable names on windows, the `system-agent-installer-rke2` script would receive an environment variable with the name `$env:$env:RKE2_DATA_DIR`. If we were to properly format that variable in this PR, other operations done on WIndows nodes would not respect the new data-dir and would fail to find specific files on the disk.

**Review by jiaqiluo — APPROVED** [review]
Good Job! 
LGTM 💯

**Comment by HarrisonWAffel** [comment]
Thanks for the reviews, I'm going to hold off on merging this until https://github.com/rancher/wins/pull/248 and https://github.com/rancher/system-agent/pull/201 are merged. I'll bump both versions as an update to this PR

**Comment by HarrisonWAffel** [comment]
This PR has been updated based off of the changes made in https://github.com/rancher/system-agent/pull/201, which places the new `ResetFailureCountOnSystemAgentRestart` field on the `Plan` struct as opposed to the `OneTimeInstruction`. This simplifies some parts os this PR

File: pkg/capr/planner/planner.go
Comment (jakefhyde) [inline pkg/capr/planner/planner.go:854]: I hate how long this function signature is getting but I don't think there is anything we can do about it yet lol.

File: pkg/capr/common.go
Comment (jakefhyde) [inline pkg/capr/common.go:684]: IMO we should try to refactor this to use the `rkev1.EnvVar` if we can since that is what is on the control plane, not sure if its doable but that way you won't have to rewrap them everywhere.

File: pkg/apis/rke.cattle.io/v1/plan/plan.go
Comment (jakefhyde) [inline pkg/apis/rke.cattle.io/v1/plan/plan.go:82]: I'm a fan of having thename and json tag match, WDYT?

File: pkg/capr/planner/planner.go
Comment (jakefhyde) [inline pkg/capr/planner/planner.go:883]:
```suggestion
		plan.ResetFailureCountOnSystemAgentRestart = resetFailureCountOnSystemAgentRestart
```

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/apis/rke.cattle.io/v1/plan/plan.go
Comment (HarrisonWAffel) [reply pkg/apis/rke.cattle.io/v1/plan/plan.go:82]: I am too in general, but in this case I think the difference makes sense. The json tag has to match in both the system-agent repo and rancher in order for the plan to be handed off via the secret. It doesn't make sense to have the json tag include the term `system-agent` inside the system-agent repo.

File: pkg/capr/common.go
Comment (HarrisonWAffel) [reply pkg/capr/common.go:700]: Good point!

File: pkg/capr/planner/planner.go
Comment (HarrisonWAffel) [reply pkg/capr/planner/planner.go:883]: Good point, the conditional was more useful during testing

File: pkg/capr/common.go
Comment (HarrisonWAffel) [reply pkg/capr/common.go:684]: I had debated between the two, but since `WindowsInstallScript` already uses `corev1` env vars I went with that type. Right now we are only wrapping a `rkev1.EnvVar` into a `corev1.EnVar` in the `generateInstallInstruction` function when passing the agent env vars into the plan. Everywhere else we are either already working with a `corev1.EnvVar` or creating one based off of simple strings. The difference isn't a big deal to me, if we think it will be more useful to move to `rkev1.EnvVar` I can make that change

File: pkg/capr/installer/installer.go
Comment (jiaqiluo) [inline pkg/capr/installer/installer.go:215]: Can you confirm that it is safe to have  `$env:CATTLE_ROLE_NONE="true` without checking if the token is empty or not?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/capr/installer/installer.go
Comment (HarrisonWAffel) [reply pkg/capr/installer/installer.go:215]: Yes, this is fine. This was previously set [within the `token` variable along with the actual token](https://github.com/rancher/rancher/blob/17221038da5bfbc17e5cd728c50925b599941ee8/pkg/capr/installer/installer.go#L163) (for some reason), and there isn't a case where it will never not be equal to `true`. We even have [a unit test](https://github.com/rancher/rancher/blob/17221038da5bfbc17e5cd728c50925b599941ee8/pkg/capr/installer/installer_test.go#L34) to ensure that it has this value.

Comment (HarrisonWAffel) [reply pkg/capr/installer/installer.go]: Good point, to be honest I'm not sure if there is a scenario where rancher delivers this script without a token being provided, without it the node wouldn't be able to contact the Rancher API. But, just to be save I've only set it when we do have the token value

File: pkg/capr/common.go
Comment (jakefhyde) [reply pkg/capr/common.go:684]: Eh just merge it, looking at the code it's probably tech debt so no point in blocking your PR to fix it. I'll probably try to fix it at some point in the future.

**Review by jakefhyde — APPROVED** [review]

**Comment by jakefhyde** [comment]
I've approved but realistically @Oats87 should also give this a :+1: before merging.

File: pkg/capr/installer/installer.go
Comment (jiaqiluo) [inline pkg/capr/installer/installer.go:182]: Can you confirm that the value is expected to be a no-boolean value?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/capr/installer/installer.go
Comment (HarrisonWAffel) [reply pkg/capr/installer/installer.go:182]: Yes, this is supposed to be a string. The current implementation also sets it as a string 

https://github.com/rancher/rancher/blob/5fd3e0305fbd6bb82d45cd98ecb2a8c4ca42616a/pkg/capr/installer/installer.go#L163

**Review by jiaqiluo — DISMISSED** [review]
LGTM

File: pkg/controllers/capr/plansecret/plansecret.go
Comment (jakefhyde) [inline pkg/controllers/capr/plansecret/plansecret.go:107]: Can you add a comment here detailed that we will set the failed checksum, but not mark the plan as truly failed unless it is either 1. not allowed to fail or 2. has reached the max failure count

File: pkg/capr/planner/planner.go
Comment (jakefhyde) [inline pkg/capr/planner/planner.go:422]: Can you expand this comment to denote why we need a higher failure threshold for Windows workers.

File: pkg/capr/common.go
Comment (jakefhyde) [inline pkg/capr/common.go:701]:
```suggestion
    // Non-boolean variables are always treated as strings, even numbers
		format := "$env:%s=\"%s\""
		if isPlanVariable {
			format = "%s=%s"
		}
```

**Review by jakefhyde — CHANGES_REQUESTED** [review]
Couple nits

**Review by jakefhyde — APPROVED** [review]

File: pkg/capr/common.go
Comment (jiaqiluo) [inline pkg/capr/common.go:697]: I will still approve the PR as the following should not be a blocker. 

```suggestion
		// Non-boolean variables are always treated as strings,
		// even numbers
		format := "$env:%s=\"%s\""
```

**Review by jiaqiluo — APPROVED** [review]

File: pkg/capr/common.go
Comment (HarrisonWAffel) [reply pkg/capr/common.go:697]: I'll update this as a part of bumping wins and system-agent once those are cut, thanks!

**Review by jiaqiluo — APPROVED** [review]

**Comment by HarrisonWAffel** [comment]
Merging with a single approval as the latest change request has been addressed and no other major changes have been made since Jakes last approval. Both the `system-agent` and `rancher-wins` have been bumped to the latest RC versions.

**Pull request merged**
